### PR TITLE
Drop support for python 3.7

### LINKED ISSUE #6138: DEP: Drop Python 3.7 Support
Community references:
- [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html) says numpy dropped support for Python 3.7 in December 2021.
- pandas (https://github.com/pandas-dev/pandas/issues/41678)
- opendatacube is Python 3.8+ already [ref](https://github.com/opendatacube/datacube-core/pull/1190)
- django [ref](https://docs.djangoproject.com/en/4.0/releases/4.0/)
- pyproj (https://github.com/pyproj4/pyproj/issues/930)
- rioxarray (https://github.com/corteva/rioxarray/issues/451)

### PR DESCRIPTION
**Pull request by Illviljan**
This PR drops support for python 3.7 and removes a bit of compatibility code related to typing.

- [x] Closes #6138
- [x] Requires #5845
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] Can be merged once we plan to make a release after 2021-12-26. https://numpy.org/neps/nep-0029-deprecation_policy.html

### CODE REVIEW

File: .github/workflows/ci-additional.yaml [hunk 1]
```diff
@@ -40,8 +40,8 @@ jobs:
         os: ["ubuntu-latest"]
         env:
           [
-            "py37-bare-minimum",
-            "py37-min-all-deps",
+            "py38-bare-minimum",
+            "py38-min-all-deps",
             "py38-all-but-dask",
             "py38-flaky",
```
Comment (keewis) [inline .github/workflows/ci-additional.yaml:46]: should these be upgraded to `py39`? And once it is actually possible, to `py310`?

**Review by keewis — COMMENTED** [review]
according to NEP29, support for `py37` (*and* `numpy=1.19`) can now be dropped. We have a slightly different policy, though, so the minimum versions script might disagree.

File: .github/workflows/ci-additional.yaml (hunk 1 above)
Comment (Illviljan) [reply .github/workflows/ci-additional.yaml]: I agree. I added some comments to clarify this.

**Comment by keewis** [comment]
what do we do about the `typing_extensions` and `importlib-resources` dependencies? Those are `py37`-only at the moment... I guess we don't need `importlib-resources` anymore, but what about `typing_extensions`? Is it a optional dependency?

Also, there's a mention of `python=3.7` in `dask_array_ops` related to the `method` parameter of `cumreduction`. What is that about?

**Comment by Illviljan** [comment]
I think `typing_extensions` should remain available so we can immediately use the latest and greatest typing functions.
For example TypeGuard https://www.python.org/dev/peps/pep-0647/ is a good one that I think we can make use of now. I've also been testing it in #5667.

**Comment by keewis** [comment]
we're already using `TypeGuard` in `xarray.core.utils`, so I guess it has to stay as a optional dependency (i.e. I will add it back to `py38-min-all-deps`). I don't think we still need the hard dependency, though.

**Comment by Illviljan** [comment]
I think it should stay required until typing has matured more, there's too much good stuff coming out each release not to make use of it immediately.

I feel having to do these kinds of workarounds is not nice and sucks out any willpower of improving typing:
https://github.com/pydata/xarray/blob/e056cacdca55cc9d9118c830ca622ea965ebcdef/xarray/core/utils.py#L309-L336
(This Try Except is not necessary anymore if it stays a required dependency)

We also don't run mypy with minimum requirements so it's very likely we miss typing errors while going through these convoluted workaround paths.

Here's a few more examples from eariler PRs:

#5678 - The necessary workaround if it's not required:
https://github.com/Illviljan/xarray/blob/c11aac2f00edf3584d1eec508fc850fb8f17e084/xarray/core/options.py#L4-L70+

#5871 - Just a simple if check if it's required:
https://github.com/dcherian/xarray/blob/59f8586a747b1f0a7b5e42e0ee186a32a3a36819/xarray/core/_reductions.py#L10-L13

**Comment by keewis** [comment]
Well, I can understand that: it's usually the same with other actively developed libraries.

My point was actually that the changes I applied shouldn't have changed the dependencies: `typing_extensions` was restricted to `python < 3.8` before, so it was only installed on `python=3.7` (unless it was installed using `conda` / from `conda-forge`, which doesn't support this kind of restriction).

**Comment by max-sixty** [comment]
I would vote to have `typing_extensions` as a dependency if it's helpful. At one point xarray-beam couldn't use the very latest version, but as long as we're OK with a version ~6 months old then I think we should be fine having it. Agree with @Illviljan re workarounds not being fun!

**Comment by keewis** [comment]
that's fine with me, but that would be a new (hard) dependency for `python>=3.8` so I think we should make that a new PR

**Comment by keewis** [comment]
should we merge this now and sort the typing out in a new PR?

**Comment by Illviljan** [comment]
Yes, let's merge.

**Pull request merged**